product_contract/models/sale_order_line.py

```diff
@@ -0,0 +1,279 @@
+            ("quarterly", "Quarter(s)"),
+            ("semesterly", "Semester(s)"),
+            ("yearly", "Year(s)"),
+        ],
+        default="monthly",
+        string="Invoice Every",
+        copy=False,
+    )
+    recurring_invoicing_type = fields.Selection(
+        [("pre-paid", "Pre-paid"), ("post-paid", "Post-paid")],
+        default="pre-paid",
+        string="Invoicing type",
+        help="Specify if process date is 'from' or 'to' invoicing date",
+        copy=False,
+    )
+    date_start = fields.Date()
+    date_end = fields.Date()
+
+    contract_line_id = fields.Many2one(
+        comodel_name="contract.line",
+        string="Contract Line to replace",
+        required=False,
+        copy=False,
+    )
+    is_auto_renew = fields.Boolean(
+        string="Auto Renew",
+        compute="_compute_auto_renew",
+        default=False,
+        store=True,
+        readonly=False,
+    )
+    auto_renew_interval = fields.Integer(
+        default=1,
+        string="Renew Every",
+        compute="_compute_auto_renew",
+        store=True,
+        readonly=False,
+        help="Renew every (Days/Week/Month/Year)",
+    )
+    auto_renew_rule_type = fields.Selection(
+        [
+            ("daily", "Day(s)"),
+            ("weekly", "Week(s)"),
+            ("monthly", "Month(s)"),
+            ("yearly", "Year(s)"),
+        ],
+        default="yearly",
+        compute="_compute_auto_renew",
+        store=True,
+        readonly=False,
+        string="Renewal type",
+        help="Specify Interval for automatic renewal.",
+    )
+
+    @api.constrains("contract_id")
+    def _check_contact_is_not_terminated(self):
+        for rec in self:
+            if (
+                rec.order_id.state not in ("sale", "done", "cancel")
+                and rec.contract_id.is_terminated
+            ):
+                raise ValidationError(
+                    _("You can't upsell or downsell a terminated contract")
+                )
+
+    @api.depends("product_id", "order_id.company_id")
+    def _compute_contract_template_id(self):
+        for rec in self:
+            rec.contract_template_id = rec.product_id.with_company(
+                rec.order_id.company_id
+            ).property_contract_template_id
+
+    def _get_auto_renew_rule_type(self):
+        """monthly last day don't make sense for auto_renew_rule_type"""
+        self.ensure_one()
+        if self.recurring_rule_type == "monthlylastday":
+            return "monthly"
+        return self.recurring_rule_type
+
+    def _get_date_end(self):
+        self.ensure_one()
+        contract_line_model = self.env["contract.line"]
+        date_end = (
+            self.date_start
+            + contract_line_model.get_relative_delta(
+                self._get_auto_renew_rule_type(),
+                int(self.product_uom_qty),
+            )
+            - relativedelta(days=1)
+        )
+        return date_end
+
+    @api.depends("product_id")
```

Comment:
@rousseldenis, @sbejaoui, @pedrobaeza, @sbidoul  maybe you could help me here please. I am trying to use this module in e-shop. However when I purchase the product, Contract is created but all the relevant fields are empty. What I found out is that this method `_compute_auto_renew` is not called in such scenario. I'm scratching my head on why is it never called and how to solve this in some good and elegant way. Any help / advice is really appreciated.

Reply:
Hi @Rad0van ,
I think the issue is that all the fields computed by this method have a default value.
What happens at record creation is that get assigned their default value so they don't need to be computed;
The computation method triggers-in, correctly, next time the observed field (`product_id`) changes.

This issue doesn't happen when a user manually creates a Sale Order because the computation method is immediately triggered when `product_id` changes (as the computation method is also part of the `onchange` workflow).

Besides causing this issue, a default value on a computed field doesn't make much sense in my opinion. I suggest you remove all defaults from field definition and maybe add them in the computation method itself.

Reply:
> 

Hi @espo-tony ,

not really sure I follow what you suggest here because the issue is as you point out that when Sale Order is created directly, the computation method is not called. I have created a hacky solution by creating glue module between product_contract and website_sale where I basically reproduce the whole `_compute_auto_renew` method except for setting the quantity. It goes like this:

```
class SaleOrder(models.Model):
    _inherit = "sale.order"

    def _cart_update_order_line(self, product_id, quantity, order_line, **kwargs):
        """
        This ensures that when contract product is put into basket in e-commerce
        all necessary values are set
        """
        order_line = super()._cart_update_order_line(
            product_id, quantity, order_line, **kwargs
        )

        """
        This is (apart from commented out line) 1:1 copy of _compute_auto_renew method
        from product_contract module to ensure we get the defaults configured
        """
        if order_line.product_id.is_contract:
            # order_line.product_uom_qty = order_line.product_id.default_qty
            order_line.recurring_rule_type = order_line.product_id.recurring_rule_type
            order_line.recurring_invoicing_type = (
                order_line.product_id.recurring_invoicing_type
            )
            order_line.date_start = order_line.date_start or fields.Date.today()

            order_line.date_end = order_line._get_date_end()
            order_line.is_auto_renew = order_line.product_id.is_auto_renew
            if order_line.is_auto_renew:
                order_line.auto_renew_interval = (
                    order_line.product_id.auto_renew_interval
                )
                order_line.auto_renew_rule_type = (
                    order_line.product_id.auto_renew_rule_type
                )

        return order_line

```

Reply:
> Hi @Rad0van When a Sale Order is created directly, the computation method is not called **because of** the default values. If you remove them, the computation method will be called even when a Sale Order is created directly. I tested this solution locally and it worked for me.

Oh I see now! But that is quite a big change which I think should be approved by more senior people.
Also I would be interested to get to know the inner mechanics of this. I.e. why the *@api.depends* is called when the default values are not present? It only depends on *product_id* Does anyone know the explanation?